Hi all. I will use some work of **@alexisbcook** ([Titanic Tutorial](https://www.kaggle.com/alexisbcook/titanic-tutorial "Alexis Cook’s Titanic Tutorial")) and **@dansbecker** ([Intro to Machine Learning](https://www.kaggle.com/learn/intro-to-machine-learning "Dan Becker’s course")) in this notebook. My purpose is to use all columns of the training data in my model with the most reasonable imputations. I will also impute to test data.

At the end of the notebook there is a references section for some of my copy & pastes. The others can be found next to code I pasted.

In [1]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error
from sklearn.ensemble import RandomForestClassifier

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

/kaggle/input/titanic/train.csv
/kaggle/input/titanic/test.csv
/kaggle/input/titanic/gender_submission.csv


Get the data...

In [2]:
# For displaying all the rows of the data frame
pd.set_option('display.max_rows', None) # Thanks to @ https://dev.to/chanduthedev/how-to-display-all-rows-from-data-frame-using-pandas-dha

# get training data
train_data = pd.read_csv("/kaggle/input/titanic/train.csv")
train_data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [3]:
# get test data
test_data = pd.read_csv("/kaggle/input/titanic/test.csv")
test_data.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

First let's search for some new patterns on both datasets. So far we have only used "Pclass", "Sex", "SibSp" and "Parch" columns ([Titanic Tutorial](https://www.kaggle.com/alexisbcook/titanic-tutorial "Alexis Cook’s Titanic Tutorial")). 

In [4]:
# Describe training data
train_data.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [5]:
# Describe test data
test_data.describe()

PassengerId      Pclass         Age       SibSp       Parch        Fare
count   418.000000  418.000000  332.000000  418.000000  418.000000  417.000000
mean   1100.500000    2.265550   30.272590    0.447368    0.392344   35.627188
std     120.810458    0.841838   14.181209    0.896760    0.981429   55.907576
min     892.000000    1.000000    0.170000    0.000000    0.000000    0.000000
25%     996.250000    1.000000   21.000000    0.000000    0.000000    7.895800
50%    1100.500000    3.000000   27.000000    0.000000    0.000000   14.454200
75%    1204.750000    3.000000   39.000000    1.000000    0.000000   31.500000
max    1309.000000    3.000000   76.000000    8.000000    9.000000  512.329200

In [6]:
# Check correlation in training data
train_data.corr(method='pearson')

PassengerId  Survived    Pclass       Age     SibSp     Parch  \
PassengerId     1.000000 -0.005007 -0.035144  0.036847 -0.057527 -0.001652   
Survived       -0.005007  1.000000 -0.338481 -0.077221 -0.035322  0.081629   
Pclass         -0.035144 -0.338481  1.000000 -0.369226  0.083081  0.018443   
Age             0.036847 -0.077221 -0.369226  1.000000 -0.308247 -0.189119   
SibSp          -0.057527 -0.035322  0.083081 -0.308247  1.000000  0.414838   
Parch          -0.001652  0.081629  0.018443 -0.189119  0.414838  1.000000   
Fare            0.012658  0.257307 -0.549500  0.096067  0.159651  0.216225   

                 Fare  
PassengerId  0.012658  
Survived     0.257307  
Pclass      -0.549500  
Age          0.096067  
SibSp        0.159651  
Parch        0.216225  
Fare         1.000000

Let' search for null values...

In [7]:
# Look null values for train data
train_data.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [8]:
# Look null values for test data
test_data.isnull().sum()

PassengerId      0
Pclass           0
Name             0
Sex              0
Age             86
SibSp            0
Parch            0
Ticket           0
Fare             1
Cabin          327
Embarked         0
dtype: int64

It seems that **'Fare'** and **'Embarked'** columns are encouraging for new patterns since they have very few null values which can be modified somehow.

Moreover correlation of 'Fare' with 'Survived' is relatively **high** (second after **'Pclass'**). This is as expected. Probably how much they paid for the tickets determined where they slept at that unholly night.

Let's see who have these missing values.

In [9]:
# There is 1 null value for 'Fare' column in test data
# https://dzone.com/articles/pandas-find-rows-where-columnfield-is-null
test_data[test_data["Fare"].isnull()]

PassengerId  Pclass                Name   Sex   Age  SibSp  Parch Ticket  \
152         1044       3  Storey, Mr. Thomas  male  60.5      0      0   3701   

     Fare Cabin Embarked  
152   NaN   NaN        S

In [10]:
# There are 2 null values for 'Embarked' column in train data
# https://dzone.com/articles/pandas-find-rows-where-columnfield-is-null
train_data[train_data["Embarked"].isnull()]

PassengerId  Survived  Pclass                                       Name  \
61            62         1       1                        Icard, Miss. Amelie   
829          830         1       1  Stone, Mrs. George Nelson (Martha Evelyn)   

        Sex   Age  SibSp  Parch  Ticket  Fare Cabin Embarked  
61   female  38.0      0      0  113572  80.0   B28      NaN  
829  female  62.0      0      0  113572  80.0   B28      NaN

If we can replace missing 'Fare' data of **"Storey, Mr. Thomas"** with a meaningful value we can use the '*Fare*' column in our model (we do not want to drop any non value since our purpose is the use all the data as much as possible).

Similarly if we can replace missing 'Embarked' data of **"Icard, Miss. Amelie"** and **"Stone Mrs. George Nelson (Martha Evelyn)"** with meaningful values we can also use the "*Embarked*" column in our model.

Let's start with **Mr. Storey**:

**Mr. Storey** embarked from Southampton. He was traveling in the third class section. Let's look who else is in the same status at datasets.

In [11]:
# Description of the people from Southampton who are traveling in the third class(training dataset)
South_3C_train_data = train_data[train_data["Embarked"] == "S"][train_data["Pclass"] == 3]
South_3C_train_data.describe()

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  


PassengerId    Survived  Pclass         Age       SibSp       Parch  \
count   353.000000  353.000000   353.0  290.000000  353.000000  353.000000   
mean    440.685552    0.189802     3.0   25.696552    0.705382    0.439093   
std     263.352933    0.392701     0.0   12.110906    1.529408    0.954863   
min       1.000000    0.000000     3.0    1.000000    0.000000    0.000000   
25%     201.000000    0.000000     3.0   19.000000    0.000000    0.000000   
50%     432.000000    0.000000     3.0   25.000000    0.000000    0.000000   
75%     668.000000    0.000000     3.0   32.000000    1.000000    0.000000   
max     889.000000    1.000000     3.0   74.000000    8.000000    6.000000   

             Fare  
count  353.000000  
mean    14.644083  
std     13.276609  
min      0.000000  
25%      7.854200  
50%      8.050000  
75%     16.100000  
max     69.550000

In [12]:
# Description of the people from Southampton who are traveling in the third class(test dataset)
South_3C_test_data = test_data[test_data["Embarked"] == "S"][test_data["Pclass"] == 3]
South_3C_test_data.describe()

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  


PassengerId  Pclass         Age       SibSp       Parch        Fare
count   142.000000   142.0  109.000000  142.000000  142.000000  141.000000
mean   1095.563380     3.0   24.051193    0.549296    0.556338   13.913030
std     118.465332     0.0   11.000006    1.274838    1.391589   12.744667
min     893.000000     3.0    0.170000    0.000000    0.000000    3.170800
25%     995.500000     3.0   19.000000    0.000000    0.000000    7.854200
50%    1092.000000     3.0   24.000000    0.000000    0.000000    8.050000
75%    1190.000000     3.0   29.000000    1.000000    0.000000   14.500000
max    1308.000000     3.0   60.500000    8.000000    9.000000   69.550000

It seems that according to the training dataset aproximately **39%** of the passengers embarked from Southampton and were travelling in the third class with an avarage fare of **£ 14.644083**.

It also seems that according to the test dataset aproximately **34%** of the passengers embarked from Southampton and were travelling in the third class with an avarage fare of **£ 13.913030**.

Note that percentage values are for third class passengers only. There may be passangers in the first and second class who are coming from Southampton

First we can use some number between these amounts (£ 13.913030 - £ 14.644083) for Mr Storey's fare. 

But before that let's check **"Icard, Miss. Amelie"** and **"Stone Mrs. George Nelson (Martha Evelyn)"** again. Although their embark location is not known it seems that they were in the same cabin (B28) and paid the same fare (£ 80.0). As can be seen easily their ticket number is also the same (113572).

So is there any other person whose ticket number is the same with Mr Storey? If yes we can conclude that they paid the same fare.

In [13]:
# Look if there is another person who has the same ticket number with Mr Storey 
train_data[train_data["Ticket"] == "3701"]

Empty DataFrame
Columns: [PassengerId, Survived, Pclass, Name, Sex, Age, SibSp, Parch, Ticket, Fare, Cabin, Embarked]
Index: []

In [14]:
# Look if there is another person who has the same ticket number with Mr Storey 
test_data[test_data["Ticket"] == "3701"]

PassengerId  Pclass                Name   Sex   Age  SibSp  Parch Ticket  \
152         1044       3  Storey, Mr. Thomas  male  60.5      0      0   3701   

     Fare Cabin Embarked  
152   NaN   NaN        S

Unfortunately no one but himself has the same ticket number with Mr Storey. May be this is because he doesn't have a cabin. Let's dive in to fares a litle bit more...

In [15]:
# Who paid £ 7.854200 in Southampton(training dataset)
train_data[train_data["Embarked"] == 'S'][train_data["Fare"] == 7.854200]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  


PassengerId  Survived  Pclass  \
14            15         0       3   
91            92         0       3   
175          176         0       3   
192          193         1       3   
281          282         0       3   
315          316         1       3   
396          397         0       3   
569          570         1       3   
623          624         0       3   
640          641         0       3   
704          705         0       3   
771          772         0       3   
833          834         0       3   

                                                Name     Sex   Age  SibSp  \
14              Vestrom, Miss. Hulda Amanda Adolfina  female  14.0      0   
91                        Andreasson, Mr. Paul Edvin    male  20.0      0   
175                           Klasen, Mr. Klas Albin    male  18.0      1   
192  Andersen-Jensen, Miss. Carla Christine Nielsine  female  19.0      1   
281                 Olsson, Mr. Nils Johan Goransson    male  28.0      0   
315                  Nilsson, Miss. Helmina Josefina  female  26.0      0   
396                              Olsson, Miss. Elina  female  31.0      0   
569                                Jonsson, Mr. Carl    male  32.0      0   
623                      Hansen, Mr. Henry Damsgaard    male  21.0      0   
640                           Jensen, Mr. Hans Peder    male  20.0      0   
704                          Hansen, Mr. Henrik Juul    male  26.0      1   
771                          Jensen, Mr. Niels Peder    male  48.0      0   
833                           Augustsson, Mr. Albert    male  23.0      0   

     Parch  Ticket    Fare Cabin Embarked  
14       0  350406  7.8542   NaN        S  
91       0  347466  7.8542   NaN        S  
175      1  350404  7.8542   NaN        S  
192      0  350046  7.8542   NaN        S  
281      0  347464  7.8542   NaN        S  
315      0  347470  7.8542   NaN        S  
396      0  350407  7.8542   NaN        S  
569      0  350417  7.8542   NaN        S  
623      0  350029  7.8542   NaN        S  
640      0  350050  7.8542   NaN        S  
704      0  350025  7.8542   NaN        S  
771      0  350047  7.8542   NaN        S  
833      0  347468  7.8542   NaN        S

In [16]:
# Who paid £ 7.854200 in Southampton(test dataset)
test_data[test_data["Embarked"] == 'S'][test_data["Fare"] == 7.854200]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  


PassengerId  Pclass                               Name     Sex   Age  \
98           990       3       Braf, Miss. Elin Ester Maria  female  20.0   
135         1027       3          Carlsson, Mr. Carl Robert    male  24.0   
157         1049       3           Lundin, Miss. Olga Elida  female  23.0   
195         1087       3  Karlsson, Mr. Julius Konrad Eugen    male  33.0   
235         1127       3             Vendel, Mr. Olof Edvin    male  20.0   
261         1153       3      Nilsson, Mr. August Ferdinand    male  21.0   
299         1191       3                Johansson, Mr. Nils    male  29.0   
318         1210       3          Jonsson, Mr. Nils Hilding    male  27.0   

     SibSp  Parch  Ticket    Fare Cabin Embarked  
98       0      0  347471  7.8542   NaN        S  
135      0      0  350409  7.8542   NaN        S  
157      0      0  347469  7.8542   NaN        S  
195      0      0  347465  7.8542   NaN        S  
235      0      0  350416  7.8542   NaN        S  
261      0      0  350410  7.8542   NaN        S  
299      0      0  347467  7.8542   NaN        S  
318      0      0  350408  7.8542   NaN        S

As can be seen easily people who paid the same fares have some resemblance in their ticket number. The tickets that cost £ 7.8542 are 6 digit numbers that start with 34 or 35. Probably they were in the same area next to each other while traveling and as a result paid the same amount.

Can we find some ticket number similar to Mr Storey?

In [17]:
# Who else has 4 digit tickets, embarked from Southampton and traveling in the third class? 
train_data[[len(x) == 4 for x in train_data["Ticket"]]][train_data["Embarked"] == "S"][train_data["Pclass"] == 3]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  


PassengerId  Survived  Pclass  \
40            41         0       3   
74            75         1       3   
103          104         0       3   
113          114         0       3   
138          139         0       3   
169          170         0       3   
176          177         0       3   
179          180         0       3   
197          198         0       3   
229          230         0       3   
271          272         1       3   
302          303         0       3   
338          339         1       3   
350          351         0       3   
402          403         0       3   
409          410         0       3   
474          475         0       3   
477          478         0       3   
483          484         1       3   
485          486         0       3   
503          504         0       3   
509          510         1       3   
597          598         0       3   
643          644         1       3   
652          653         0       3   
677          678         1       3   
682          683         0       3   
692          693         1       3   
713          714         0       3   
769          770         0       3   
810          811         0       3   
826          827         0       3   
834          835         0       3   
838          839         1       3   
876          877         0       3   
882          883         0       3   

                                               Name     Sex   Age  SibSp  \
40   Ahlin, Mrs. Johan (Johanna Persdotter Larsson)  female  40.0      1   
74                                    Bing, Mr. Lee    male  32.0      0   
103                      Johansson, Mr. Gustaf Joel    male  33.0      0   
113                         Jussila, Miss. Katriina  female  20.0      1   
138                             Osen, Mr. Olaf Elon    male  16.0      0   
169                                   Ling, Mr. Lee    male  28.0      0   
176                   Lefebre, Master. Henry Forbes    male   NaN      3   
179                             Leonard, Mr. Lionel    male  36.0      0   
197                Olsen, Mr. Karl Siegwart Andreas    male  42.0      0   
229                         Lefebre, Miss. Mathilde  female   NaN      3   
271                    Tornquist, Mr. William Henry    male  25.0      0   
302                 Johnson, Mr. William Cahoone Jr    male  19.0      0   
338                           Dahl, Mr. Karl Edwart    male  45.0      0   
350                          Odahl, Mr. Nils Martin    male  23.0      0   
402                        Jussila, Miss. Mari Aina  female  21.0      1   
409                              Lefebre, Miss. Ida  female   NaN      3   
474                     Strandberg, Miss. Ida Sofia  female  22.0      0   
477                       Braund, Mr. Lewis Richard    male  29.0      1   
483                          Turkula, Mrs. (Hedwig)  female  63.0      0   
485                          Lefebre, Miss. Jeannie  female   NaN      3   
503                  Laitinen, Miss. Kristina Sofia  female  37.0      0   
509                                  Lang, Mr. Fang    male  26.0      0   
597                             Johnson, Mr. Alfred    male  49.0      0   
643                                 Foo, Mr. Choong    male   NaN      0   
652                  Kalvik, Mr. Johannes Halvorsen    male  21.0      0   
677                         Turja, Miss. Anna Sofia  female  18.0      0   
682                     Olsvigen, Mr. Thor Anderson    male  20.0      0   
692                                    Lam, Mr. Ali    male   NaN      0   
713                      Larsson, Mr. August Viktor    male  29.0      0   
769                Gronnestad, Mr. Daniel Danielsen    male  32.0      0   
810                          Alexander, Mr. William    male  26.0      0   
826                                    Lam, Mr. Len    male   NaN      0   
834                          Allum, Mr. Owen George    male  18.0      0   
838           

In [18]:
# Who else has 4 digit tickets, embarked from Southampton and traveling in the third class? 
test_data[[len(x) == 4 for x in test_data["Ticket"]]][test_data["Embarked"] == "S"][test_data["Pclass"] == 3]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  


PassengerId  Pclass                              Name     Sex   Age  \
5            897       3        Svensson, Mr. Johan Cervin    male  14.0   
39           931       3                     Hee, Mr. Ling    male   NaN   
108         1000       3  Willer, Mr. Aaron (Abi Weller")"    male   NaN   
132         1024       3     Lefebre, Mrs. Frank (Frances)  female   NaN   
152         1044       3                Storey, Mr. Thomas    male  60.5   
169         1061       3      Hellstrom, Miss. Hilda Maria  female  22.0   
243         1135       3                Hyman, Mr. Abraham    male   NaN   
253         1145       3          Salander, Mr. Karl Johan    male  24.0   
357         1249       3               Lockyer, Mr. Edward    male   NaN   

     SibSp  Parch Ticket     Fare Cabin Embarked  
5        0      0   7538   9.2250   NaN        S  
39       0      0   1601  56.4958   NaN        S  
108      0      0   3410   8.7125   NaN        S  
132      0      4   4133  25.4667   NaN        S  
152      0      0   3701      NaN   NaN        S  
169      0      0   7548   8.9625   NaN        S  
243      0      0   3470   7.8875   NaN        S  
253      0      0   7266   9.3250   NaN        S  
357      0      0   1222   7.8792   NaN        S

4 digit tickets does not reveal much. The only resemblancethat can found are the tickets:
* 3470(£ 7.8875),
* 3410(£ 8.7125),
* 3474(£ 7.8875),
* 3460(£ 7.0458),

Obviously there is not much similarity other that they all start with 3. But may be we can use £ 7.8875 for Mr Storey. Let's try another option.

In [19]:
# Are there any tickets that contains the string 3701? May be Some digits in Mr Storey' ticket is missing.
train_data[['3701' in x for x in train_data['Ticket']]]

PassengerId  Survived  Pclass                                      Name  \
254          255         0       3  Rosblom, Mrs. Viktor (Helena Wilhelmina)   
424          425         0       3               Rosblom, Mr. Viktor Richard   

        Sex   Age  SibSp  Parch  Ticket     Fare Cabin Embarked  
254  female  41.0      0      2  370129  20.2125   NaN        S  
424    male  18.0      1      1  370129  20.2125   NaN        S

In [20]:
# Are there any tickets that contains the string 3701? May be Some digits in Mr Storey' ticket is missing.
test_data[['3701' in x for x in test_data['Ticket']]]

PassengerId  Pclass                         Name     Sex   Age  SibSp  \
152         1044       3           Storey, Mr. Thomas    male  60.5      0   
284         1176       3  Rosblom, Miss. Salli Helena  female   2.0      1   

     Parch  Ticket     Fare Cabin Embarked  
152      0    3701      NaN   NaN        S  
284      1  370129  20.2125   NaN        S

4 people in the datasets have string 3701 on their tickets. One of them is Mr Storey whose fare we do not know. The others are Rosblom family. Their ticket number is '370129' and paid £ 20.2125 each. May be Mr Storey's ticket was something like '370128' and was torn somehow and we lost the last 2 digit and the fare amount together. This makes sense. Also explains why Mr Storey's fare is missing.

Which one shall we use for Mr Storey's fare?
* £ 14.644083 (the avarage training data fare)?
* £ 13.913030 (the avarage test data fare)?
* £ 7.8875 (4 digit tickets)?
* £ 20.2125 (tickets contain '3701')?

Let's use £ 20.2125 for Mr Storey's fare since torn ticket theory also explains why Mr Storey's fare is missing 

In [21]:
# Insert £ 20.2125 for Mr Storey's fare
test_data.loc[152,'Fare'] = 20.2125

In [22]:
# Check Mr Storey
test_data.loc[152]

PassengerId                  1044
Pclass                          3
Name           Storey, Mr. Thomas
Sex                          male
Age                          60.5
SibSp                           0
Parch                           0
Ticket                       3701
Fare                      20.2125
Cabin                         NaN
Embarked                        S
Name: 152, dtype: object

In [23]:
# Check again
test_data.isnull().sum()

PassengerId      0
Pclass           0
Name             0
Sex              0
Age             86
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          327
Embarked         0
dtype: int64

OK. Now we can use **'Fare'** column in our model. Now it is time for **'Embarked'** column. Let's remember the missing values...

In [24]:
# There are 2 null values for 'Embarked' column in train data
# https://dzone.com/articles/pandas-find-rows-where-columnfield-is-null
train_data[train_data["Embarked"].isnull()]

PassengerId  Survived  Pclass                                       Name  \
61            62         1       1                        Icard, Miss. Amelie   
829          830         1       1  Stone, Mrs. George Nelson (Martha Evelyn)   

        Sex   Age  SibSp  Parch  Ticket  Fare Cabin Embarked  
61   female  38.0      0      0  113572  80.0   B28      NaN  
829  female  62.0      0      0  113572  80.0   B28      NaN

**Miss Icard's** and **Mrs Stone's** embarking location is missing. Let's start with **'Fare'**. Is there anybody else who gave **£ 80.0** for their ticket?

In [25]:
# Who paid £ 80 in training data 
train_data[train_data["Fare"] == 80]

PassengerId  Survived  Pclass                                       Name  \
61            62         1       1                        Icard, Miss. Amelie   
829          830         1       1  Stone, Mrs. George Nelson (Martha Evelyn)   

        Sex   Age  SibSp  Parch  Ticket  Fare Cabin Embarked  
61   female  38.0      0      0  113572  80.0   B28      NaN  
829  female  62.0      0      0  113572  80.0   B28      NaN

In [26]:
# Who paid £ 80 in test data
test_data[test_data["Fare"] == 80]

Empty DataFrame
Columns: [PassengerId, Pclass, Name, Sex, Age, SibSp, Parch, Ticket, Fare, Cabin, Embarked]
Index: []

No one else paid that amount of money. Let's check then who else paid **£ 70.0** or more.

In [27]:
# People who paid more than £ 70 in training data
train_data[train_data["Fare"] > 70]

PassengerId  Survived  Pclass  \
1              2         1       1   
27            28         0       1   
31            32         1       1   
34            35         0       1   
52            53         1       1   
61            62         1       1   
62            63         0       1   
72            73         0       2   
88            89         1       1   
102          103         0       1   
118          119         0       1   
120          121         0       2   
124          125         0       1   
139          140         0       1   
195          196         1       1   
215          216         1       1   
218          219         1       1   
224          225         1       1   
230          231         1       1   
245          246         0       1   
256          257         1       1   
257          258         1       1   
258          259         1       1   
262          263         0       1   
268          269         1       1   
269          270         1       1   
275          276         1       1   
290          291         1       1   
291          292         1       1   
297          298         0       1   
299          300         1       1   
305          306         1       1   
306          307         1       1   
307          308         1       1   
310          311         1       1   
311          312         1       1   
318          319         1       1   
319          320         1       1   
325          326         1       1   
332          333         0       1   
334          335         1       1   
337          338         1       1   
341          342         1       1   
366          367         1       1   
373          374         0       1   
375          376         1       1   
377          378         0       1   
380          381         1       1   
385          386         0       2   
390          391         1       1   
393          394         1       1   
412          413         1       1   
435          436         1       1   
438          439         0       1   
445          446         1       1   
453          454         1       1   
484          485         1       1   
486          487         1       1   
496          497         1       1   
498          499         0       1   
504          505         1       1   
505          506         0       1   
520          521         1       1   
527          528         0       1   
537          538         1       1   
540          541         1       1   
544          545         0       1   
550          551         1       1   
557          558         0       1   
558          559         1       1   
581          582         1       1   
585          586         1       1   
587          588         1       1   
591          592         1       1   
609          610         1       1   
627          628         1       1   
645          646         1       1   
655          656         0       2   
659          660         0       1   
660          661         1       1   
665          666         0       2   
679          680         1       1   
681          682         1       1   
689          690         1       1   
698          699         0       1   
700          701         1       1   
708          709         1       1   
716          717         1       1   
730          731         1       1   
737          738         1       1   
741          742         0       1   
742          743         1       1   
745          746         0       1   
759          760         1       1   
763          764         1       1   
765          766         1       1   
779          780         1       1   
789          790         0       1   
802          803         1       1   
820          821         1       1   
829          830         1       1   
835          836         1       1   
849          850         1       1   
856          857         1       1   
879          88

In [28]:
# People who paid more than £ 70 in test data
test_data[test_data["Fare"] > 70]

PassengerId  Pclass                                               Name  \
12           904       1      Snyder, Mrs. John Pillsbury (Nelle Stevenson)   
24           916       1    Ryerson, Mrs. Arthur Larned (Emily Maria Borie)   
48           940       1    Bucknell, Mrs. William Robert (Emma Eliza Ward)   
53           945       1                         Fortune, Miss. Ethel Flora   
59           951       1                        Chaudanson, Miss. Victorine   
64           956       1                        Ryerson, Master. John Borie   
69           961       1                Fortune, Mrs. Mark (Mary McDougald)   
74           966       1                               Geiger, Miss. Amalie   
75           967       1                                 Keeping, Mr. Edwin   
81           973       1                                 Straus, Mr. Isidor   
96           988       1  Cavendish, Mrs. Tyrell William (Julia Florence...   
114         1006       1             Straus, Mrs. Isidor (Rosalie Ida Blun)   
118         1010       1                               Beattie, Mr. Thomson   
141         1033       1                               Daniels, Miss. Sarah   
142         1034       1                         Ryerson, Mr. Arthur Larned   
150         1042       1              Earnshaw, Mrs. Boulton (Olive Potter)   
156         1048       1                                  Bird, Miss. Ellen   
179         1071       1  Compton, Mrs. Alexander Taylor (Mary Eliza Ing...   
181         1073       1                   Compton, Mr. Alexander Taylor Jr   
184         1076       1  Douglas, Mrs. Frederick Charles (Mary Helene B...   
196         1088       1                    Spedden, Master. Robert Douglas   
202         1094       1                             Astor, Col. John Jacob   
212         1104       2                          Deacon, Mr. Percy William   
217         1109       1                           Wick, Mr. George Dennick   
218         1110       1       Widener, Mrs. George Dunton (Eleanor Elkins)   
234         1126       1                          Cumings, Mr. John Bradley   
236         1128       1                           Warren, Mr. Frank Manley   
239         1131       1        Douglas, Mrs. Walter Donald (Mahala Dutton)   
242         1134       1                       Spedden, Mr. Frederic Oakley   
252         1144       1                           Clark, Mr. Walter Miller   
270         1162       1                       McCaffry, Mr. Thomas Francis   
272         1164       1      Clark, Mrs. Walter Miller (Virginia McDowell)   
287         1179       1                         Snyder, Mr. John Pillsbury   
293         1185       1                              Dodge, Dr. Washington   
306         1198       1               Allison, Mr. Hudson Joshua Creighton   
308         1200       1                         Hays, Mr. Charles Melville   
314         1206       1              White, Mrs. John Stuart (Ella Holmes)   
316         1208       1                      Spencer, Mr. William Augustus   
324         1216       1                             Kreuchen, Miss. Emilie   
327         1219       1        Rosenshine, Mr. George (Mr George Thorne")"   
343         1235       1  Cardeza, Mrs. James Warburton Martinez (Charlo...   
352         1244       2                                Dibden, Mr. William   
371         1263       1                          Wilson, Miss. Helen Alice   
374         1266       1              Dodge, Mrs. Washington (Ruth Vidaver)   
375         1267       1                           Bowen, Miss. Grace Scott   
390         1282       1                         Payne, Mr. Vivian Ponsonby   
397         1289       1  Frolicher-Stehli, Mrs. Maxmillian (Margaretha ...   
400         1292       1                            Bonnell, Miss. Caroline   
407         1299       1                         Widener, Mr. George Dunton   
411         1303       1    Minahan, Mrs. William Edwar

 **'Fare'** does not reveal much. Let's try a different approach. **Miss Icard's** and **Mrs Stone's** Cabin number is **B28**. Let's find the cabins with **'B'**... 

In [29]:
# people in Cabins starting with 'B' in training data
train_data[['B' in str(x) for x in train_data['Cabin']]]

PassengerId  Survived  Pclass  \
31            32         1       1   
54            55         0       1   
61            62         1       1   
118          119         0       1   
139          140         0       1   
170          171         0       1   
194          195         1       1   
195          196         1       1   
257          258         1       1   
263          264         0       1   
291          292         1       1   
299          300         1       1   
311          312         1       1   
329          330         1       1   
369          370         1       1   
390          391         1       1   
435          436         1       1   
484          485         1       1   
487          488         0       1   
504          505         1       1   
520          521         1       1   
523          524         1       1   
536          537         0       1   
539          540         1       1   
540          541         1       1   
587          588         1       1   
632          633         1       1   
641          642         1       1   
671          672         0       1   
679          680         1       1   
689          690         1       1   
690          691         1       1   
730          731         1       1   
737          738         1       1   
742          743         1       1   
745          746         0       1   
759          760         1       1   
763          764         1       1   
779          780         1       1   
781          782         1       1   
789          790         0       1   
802          803         1       1   
815          816         0       1   
820          821         1       1   
829          830         1       1   
872          873         0       1   
887          888         1       1   

                                                  Name     Sex   Age  SibSp  \
31      Spencer, Mrs. William Augustus (Marie Eugenie)  female   NaN      1   
54                      Ostby, Mr. Engelhart Cornelius    male  65.0      0   
61                                 Icard, Miss. Amelie  female  38.0      0   
118                           Baxter, Mr. Quigg Edmond    male  24.0      0   
139                                 Giglio, Mr. Victor    male  24.0      0   
170                          Van der hoef, Mr. Wyckoff    male  61.0      0   
194          Brown, Mrs. James Joseph (Margaret Tobin)  female  44.0      0   
195                               Lurette, Miss. Elise  female  58.0      0   
257                               Cherry, Miss. Gladys  female  30.0      0   
263                              Harrison, Mr. William    male  40.0      0   
291            Bishop, Mrs. Dickinson H (Helen Walton)  female  19.0      1   
299    Baxter, Mrs. James (Helene DeLaudeniere Chaput)  female  50.0      0   
311                         Ryerson, Miss. Emily Borie  female  18.0      2   
329                       Hippach, Miss. Jean Gertrude  female  16.0      0   
369                      Aubart, Mme. Leontine Pauline  female  24.0      0   
390                         Carter, Mr. William Ernest    male  36.0      1   
435                          Carter, Miss. Lucile Polk  female  14.0      1   
484                            Bishop, Mr. Dickinson H    male  25.0      1   
487                            Kent, Mr. Edward Austin    male  58.0      0   
504                              Maioni, Miss. Roberta  female  16.0      0   
520                              Perreault, Miss. Anne  female  30.0      0   
523    Hippach, Mrs. Louis Albert (Ida Sophia Fischer)  female  44.0      0   
536                  Butt, Major. Archibald Willingham    male  45.0      0   
539                 Frolicher, Miss. Hedwig Margaritha  female  22.0      0   
540                            Crosby, Miss. Harriet R  female  36.0      0   
587                   Frolicher-Stehli, Mr. Maxmillian    male  60.0      1   
632                          Stahelin-Maeglin, 

In [30]:
# people in Cabins starting with 'B' in test data
test_data[['B' in str(x) for x in test_data['Cabin']]]

PassengerId  Pclass                                               Name  \
12           904       1      Snyder, Mrs. John Pillsbury (Nelle Stevenson)   
24           916       1    Ryerson, Mrs. Arthur Larned (Emily Maria Borie)   
26           918       1                       Ostby, Miss. Helene Ragnhild   
59           951       1                        Chaudanson, Miss. Victorine   
64           956       1                        Ryerson, Master. John Borie   
92           984       1               Davidson, Mrs. Thornton (Orian Hays)   
142         1034       1                         Ryerson, Mr. Arthur Larned   
166         1058       1                                 Brandeis, Mr. Emil   
184         1076       1  Douglas, Mrs. Frederick Charles (Mary Helene B...   
215         1107       1                              Head, Mr. Christopher   
287         1179       1                         Snyder, Mr. John Pillsbury   
305         1197       1  Crosby, Mrs. Edward Gifford (Catherine Elizabe...   
308         1200       1                         Hays, Mr. Charles Melville   
316         1208       1                      Spencer, Mr. William Augustus   
343         1235       1  Cardeza, Mrs. James Warburton Martinez (Charlo...   
372         1264       1                            Ismay, Mr. Joseph Bruce   
390         1282       1                         Payne, Mr. Vivian Ponsonby   
397         1289       1  Frolicher-Stehli, Mrs. Maxmillian (Margaretha ...   

        Sex   Age  SibSp  Parch      Ticket      Fare            Cabin  \
12   female  23.0      1      0       21228   82.2667              B45   
24   female  48.0      1      3    PC 17608  262.3750  B57 B59 B63 B66   
26   female  22.0      0      1      113509   61.9792              B36   
59   female  36.0      0      0    PC 17608  262.3750              B61   
64     male  13.0      2      2    PC 17608  262.3750  B57 B59 B63 B66   
92   female  27.0      1      2  F.C. 12750   52.0000              B71   
142    male  61.0      1      3    PC 17608  262.3750  B57 B59 B63 B66   
166    male  48.0      0      0    PC 17591   50.4958              B10   
184  female  27.0      1      1    PC 17558  247.5208          B58 B60   
215    male  42.0      0      0      113038   42.5000              B11   
287    male  24.0      1      0       21228   82.2667              B45   
305  female  64.0      1      1      112901   26.5500              B26   
308    male  55.0      1      1       12749   93.5000              B69   
316    male  57.0      1      0    PC 17569  146.5208              B78   
343  female  58.0      0      1    PC 17755  512.3292      B51 B53 B55   
372    male  49.0      0      0      112058    0.0000      B52 B54 B56   
390    male  23.0      0      0       12749   93.5000              B24   
397  female  48.0      1      1       13567   79.2000              B41   

    Embarked  
12         S  
24         C  
26         C  
59         C  
64         C  
92         S  
142        C  
166        C  
184        C  
215        S  
287        S  
305        S  
308        S  
316        C  
343        C  
372        S  
390        S  
397        C

It's seen that **B**'s are either from **'Cherbourg'** or **'Southampton'**. What about **'B2'**? Remember that their Cabin is **B28**.

In [31]:
# people in Cabins starting with 'B2' in training data
train_data[['B2' in str(x) for x in train_data['Cabin']]]

PassengerId  Survived  Pclass                                       Name  \
61            62         1       1                        Icard, Miss. Amelie   
540          541         1       1                    Crosby, Miss. Harriet R   
690          691         1       1                    Dick, Mr. Albert Adrian   
745          746         0       1               Crosby, Capt. Edward Gifford   
781          782         1       1  Dick, Mrs. Albert Adrian (Vera Gillespie)   
829          830         1       1  Stone, Mrs. George Nelson (Martha Evelyn)   

        Sex   Age  SibSp  Parch     Ticket  Fare Cabin Embarked  
61   female  38.0      0      0     113572  80.0   B28      NaN  
540  female  36.0      0      2  WE/P 5735  71.0   B22        S  
690    male  31.0      1      0      17474  57.0   B20        S  
745    male  70.0      1      1  WE/P 5735  71.0   B22        S  
781  female  17.0      1      0      17474  57.0   B20        S  
829  female  62.0      0      0     113572  80.0   B28      NaN

In [32]:
test_data[['B2' in str(x) for x in test_data['Cabin']]]

PassengerId  Pclass                                               Name  \
305         1197       1  Crosby, Mrs. Edward Gifford (Catherine Elizabe...   
390         1282       1                         Payne, Mr. Vivian Ponsonby   

        Sex   Age  SibSp  Parch  Ticket   Fare Cabin Embarked  
305  female  64.0      1      1  112901  26.55   B26        S  
390    male  23.0      0      0   12749  93.50   B24        S

It's seen that cabins starting with **'B2'** are from **'Southampton'**. So let's use **'S'** for **Miss Icard** and **Mrs Stone**.

In [33]:
# Miss Icard
train_data.loc[61, 'Embarked'] = 'S'
# Mrs Stone
train_data.loc[829, 'Embarked'] = 'S'

In [34]:
# Check Miss Icard
train_data.loc[61]

PassengerId                     62
Survived                         1
Pclass                           1
Name           Icard, Miss. Amelie
Sex                         female
Age                             38
SibSp                            0
Parch                            0
Ticket                      113572
Fare                            80
Cabin                          B28
Embarked                         S
Name: 61, dtype: object

In [35]:
# Check Mrs Stone
train_data.loc[829]

PassengerId                                          830
Survived                                               1
Pclass                                                 1
Name           Stone, Mrs. George Nelson (Martha Evelyn)
Sex                                               female
Age                                                   62
SibSp                                                  0
Parch                                                  0
Ticket                                            113572
Fare                                                  80
Cabin                                                B28
Embarked                                               S
Name: 829, dtype: object

In [36]:
# Check again
train_data.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         0
dtype: int64

OK. Now we can use **'Embarked'** column in our model.

In [37]:
from sklearn.ensemble import RandomForestClassifier

y = train_data["Survived"]

features = ["Pclass", "Sex", "SibSp", "Parch", "Fare", "Embarked"]
X = pd.get_dummies(train_data[features])
X_test = pd.get_dummies(test_data[features])

model = RandomForestClassifier(n_estimators=100, max_depth=7, random_state=1)
model.fit(X, y)
predictions = model.predict(X_test)

output = pd.DataFrame({'PassengerId': test_data.PassengerId, 'Survived': predictions})
output.to_csv('my_submission.csv', index=False)
print("Your submission was successfully saved!")

Your submission was successfully saved!


**References**

Thanks to **@alexisbcook** and  **@dansbecker**

https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#links
https://stackoverflow.com/questions/13842088/set-value-for-particular-cell-in-pandas-dataframe-using-index thanks to @Yash